src/Compilers/Core/Portable/Symbols/Nullability.cs

```diff
@@ -0,0 +1,107 @@
+    /// have a nullability of <see cref="NonNull"/>, and nullable value types will always
+    /// have a nullability of <see cref="MaybeNull"/>.
+    /// </para>
+    /// <para>
+    /// Unconstrained type parameters are a unique superposition of both <see cref="NonNull"/>
+    /// and <see cref="MaybeNull"/>. They must be checked for <see langword="null"/> before
+    /// being dereferenced, but are treated as <see cref="NonNull"/> for the purposes of
+    /// assignment and return type warnings. As an example:
+    /// <code>
+    /// T M&lt;T&gt;(T t1)
+    /// {
+    ///     t1.ToString(); // Warning reported for dereferencing t1
+    ///     t1?.ToString(); // No warning reported for using ?. operator
+    ///     T t2 = default(T2); // Warning reported for assigning possibly null into t2
+    ///     return t2; // Warning reported for returning possibly null
+    /// }
+    /// </code>
+    /// This is because both <code>string</code> and <code>string?</code> can be substituted
+    /// for <code>T</code>, so the user must account for the strictest possible scenario
+    /// in all cases.
+    /// </para>
+    /// </remarks>
+    public enum Nullability
+    {
+        /// <summary>
+        /// There is no information on the current nullable state.
+        /// </summary>
+        /// <remarks>
+        /// This is used for legacy code scenarios, where a legacy API that does
+        /// not provide nullability information is being used. Expressions that
+        /// have unknown nullability are also referred to as oblivious expressions,
+        /// and they generally do not provide warnings when converting to 
+        /// <see langword="null"/>, or when dereferencing them.
+        /// </remarks>
+        Unknown = 0,
+        /// <summary>
+        /// The variable is either known or declared to be non-null.
+        /// </summary>
+        /// <remarks>
+        /// It is possible for a variable to have a declared nullability of NonNull,
+        /// but to then be inferred to be null later. As an example:
+        /// <code>
+        /// object o = null; // Warning for assigning null to a NonNull object
+        /// o.ToString(); // Warning for dereferencing null
+        /// </code>
+        /// In this example, the <see cref="ILocalSymbol.DeclaredNullability"/> will 
+        /// be <see cref="NonNull"/>, but the <see cref="TypeInfo.Nullability"/> of
+        /// <code>o</code> in the <code>o.ToString()</code> call will be 
+        /// <see cref="MaybeNull"/>, because the compiler tracks null state regardless
+        /// of the declared nullability.
+        /// </remarks>
+        NonNull = 1,
+        /// <summary>
+        /// The variable is either known or declared to possibly be null.
+        /// </summary>
+        /// <remarks>
+        /// Nullable value types will always be considered to be <see cref="MaybeNull"/>.
+        /// </remarks>
```

Comment:
I don't think this is correct. Didn't we decide to track nullability for nullable value types? If so, they can possibly be non-null.

Reply:
Currently, it does not, and it's relatively unlikely we'll get to this for preview one. I'm going to leave this as is for now, and I added it to the issue for tracking these types: [https://github.com/dotnet/roslyn/issues/29817](https://github.com/dotnet/roslyn/issues/29817)

---
In reply to: [217118070](https://github.com/dotnet/roslyn/pull/28860#discussion_r217118070) [](ancestors = 217118070)